### ISSUE TITLE
Lint rule for `typedef Foo<T> = Function<T>(T x)`

### ISSUE THREAD

**Issue by chalin** [dart-lang/linter#1027]
From https://github.com/dart-lang/sdk/issues/33165#issuecomment-397356116:

Consider adding a lint rule for the following situation:

```dart
typedef Foo<T> = Function<T>(T x);
```

cc @eernstg @leafpetersen @kevmoo @kwalrath 

_Maybe_ we also need a lint for the following case, but I'd consider it a distinct issue:

```dart
typedef Foo<T> = Function();
```

(We currently don't complain about `class C<T> {}` so maybe we shouldn't complain about a typedef that doesn't use it type argument.)

**Comment by chalin** [comment]
More generally, I guess that the rule would catch

```dart
typedef Foo<T1,...,Tn> = U0 Function<S1,...,Sk,Ti,Sk+1,...Sm>(U1,...,Uq);
```

where one of the U<sub>i</sub> is T<sub>j</sub> for some _j_ between 1 and _n_.

**Comment by srawlins** [comment]
There is now an `avoid_shadowing_type_parameters` lint rule, which should (but does not) catch this, correct?

I don't understand your general rule, @chalin, from [above](https://github.com/dart-lang/linter/issues/1027#issuecomment-397367443), but that <code>T<sub>i</sub></code> in the type parameters on `Function` should be caught, right?

**Comment by srawlins** [comment]
Dart 2.1.0 ships with linter 0.1.71, but avoid_shadowing_type_parameters is introduced in linter 0.1.72. https://github.com/dart-lang/linter/blob/0.1.76/CHANGELOG.md#0172

Did I just address this issue entirely here then? https://github.com/dart-lang/linter/commit/52cef3a7f755f111ccbe0bf6334b24cd3023fedb

**Comment by chalin** [comment]
When will that fix be released?

**Comment by srawlins** [comment]
It was released in 0.1.77: https://github.com/dart-lang/linter/blob/master/CHANGELOG.md#0177

**Comment by chalin** [comment]
Which SDK would contain 0.1.77?

**Comment by srawlins** [comment]
linter 0.1.78 is in SDK 2.1.1-dev.1.0: https://github.com/dart-lang/sdk/blob/2.1.1-dev.1.0/DEPS

**Comment by pq** [comment]
Sam beat me to it! 👍
